fix(ci): Improve workflow efficiency by optimizing runner usage and eliminating duplicate runs

jackluo923 opened the pull request:
## Description

Improve CI workflow efficiency with two changes:

### 1. Use GitHub-hosted runner for lightweight path filtering job
The `filter-relevant-changes` job only performs git operations to detect changed paths. Moving it to a GitHub-hosted runner avoids consuming self-hosted runner resources needed for heavier build jobs.

### 2. Use skip-duplicate-actions to prevent duplicate workflow runs
When pushing to a PR branch, both `push` and `pull_request` events trigger, causing duplicate workflow runs. This PR adds [fkirc/skip-duplicate-actions](https://github.com/fkirc/skip-duplicate-actions) to detect and skip duplicate runs.

**How it works:**
- When both `push` and `pull_request` trigger on the same commit, the duplicate is skipped
- Scheduled builds and manual triggers always run (`do_not_skip`)
- Downstream jobs are skipped automatically when `filter-relevant-changes` is skipped

**Why this action:**
- Used by 9,234+ repositories including [haskell-language-server](https://github.com/haskell/haskell-language-server), [CNCF TOC](https://github.com/cncf/toc), and [MegaLinter](https://github.com/oxsecurity/megalinter)
- Note: step-security/skip-duplicate-actions is a fork that requires a paid subscription; we use the original fkirc version which is free

## Validation performed

- Verified YAML syntax
- Reviewed fkirc/skip-duplicate-actions documentation


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **Chores**
  * Added a reusable check to detect and skip duplicate workflow runs, reducing unnecessary builds.
  * Made downstream pipeline steps conditional so jobs run only when appropriate via the new skip output.
  * Reworked runner configuration to prefer self-hosted when available with a GitHub-hosted fallback for reliability.
  * Clarified workflow comments and alignment for maintainability.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1212854499651194

jackluo923 committed:
fix(ci): use GitHub-hosted runner for lightweight path filtering job

The `filter-relevant-changes` job only performs git operations to detect
changed paths. Using a GitHub-hosted runner avoids consuming self-hosted
runner resources needed for heavier build jobs.

jackluo923 committed:
fix(ci): limit push event trigger to main branch only

Prevents duplicate workflow runs when pushing to PR branches. PRs are
tested via the pull_request event; the push event is only needed for
testing commits merged to main.

junhaoliao commented on .github/workflows/clp-artifact-build.yaml:
```diff
@@ -8,6 +8,8 @@ on:
       - "docs/**"
   push:
+    branches:
+      - "main"
```
@kirkrodrigues is this something you mentioned that we want to avoid? though i vaguely remember seeing something similar in our workflows previously

jackluo923 replied:
Yah. CI only triggers when pushed to their fork's main branch after this change if they don't submit a PR. Otherwise, if they PR via their own branch into our repo, our CI would be triggered. I can revert this if that's not something we want.

jackluo923 committed:
fix(ci): use skip-duplicate-actions to prevent duplicate workflow runs

Replace the push branch restriction with step-security/skip-duplicate-actions
to handle duplicate runs when both push and pull_request events trigger.

This action is:
- Used by 9,234+ repositories including haskell-language-server, CNCF TOC,
  and MegaLinter
- A security-hardened drop-in replacement for fkirc/skip-duplicate-actions
- More flexible than limiting push to main branch only

jackluo923 committed:
Merge branch 'main' into ci

jackluo923 committed:
fix(ci): switch to fkirc/skip-duplicate-actions

step-security version requires a paid subscription.

jackluo923 committed:
fix(ci): use double quotes for yamllint compliance

jackluo923 committed:
Merge branch 'main' into ci

jackluo923 committed:
refactor(ci): extract skip-duplicate-check to reusable workflow

Create a reusable workflow for skip-duplicate-check that can be used by
other workflows in the future.

junhaoliao reviewed (changes requested):
lgtm. mostly just nitpicking except that one request about clarification on `concurrency:` conflicts
